index.js

```diff
@@ -1286,12 +1288,25 @@
   //. false
   //. ```
   {
-    //  $pairs :: Array (Array2 Any Any)
-    const $pairs = [];
+    // structEquals :: Object -> Object -> Boolean
+    function structEquals(x, y) {
+      const keys = sortedKeys (x);
+      return Z.setoidEquals (keys, sortedKeys (y)) &&
+             keys.every (k => Z.equals (x[k], y[k]));
+    }
 
     Z.equals = (x, y) => {
       if (!(sameType (x, y))) return false;
+      return Z.Setoid.test (x) ? Z.setoidEquals (x, y) : structEquals (x, y);
+    };
+  }
+
+  // setoidEquals :: Setoid a => (a, a) -> Boolean
+  {
+    //  $pairs :: Array (Array2 Any Any)
+    const $pairs = [];
 
+    Z.setoidEquals = (x, y) => {
```

Comment:
I placed the function on the exports (without documentation) for two reasons:

- Consistency with other parts of the code.
- I'd like it to be a public function eventually; I believe Sanctuary could use it instead of the unprincipled equality function, since Sanctuary will have already checked that input types are Setoids of the same type all the way down.

Reply:
I agree with exporting this function. ADT libraries could use `Z.setoidEquals` in their `fantasy-land/equals` implementations, while `Z.equals` would remain available for use in test suites.